### PR TITLE
Support DSR with IPIP in CNI mode

### PR DESCRIPTION
**Pull request by Jiang1155**
```release-note
Add support for DSR using IPIP. The switches will forward the IPIP packets via a fast path.
```

### CODE REVIEW

File: bpf/include/bpf/ctx/xdp.h [hunk 1]
```diff
@@ -209,8 +209,10 @@ ctx_adjust_hroom(struct xdp_md *ctx, const __s32 len_diff, const __u32 mode,
 	void *data, *data_end;
 	int ret;
 
-	build_bug_on(len_diff <= 0 || len_diff >= 64);
+	build_bug_on(len_diff == 0 || len_diff >= 64);
 	build_bug_on(mode != BPF_ADJ_ROOM_NET);
+	if (len_diff < 0)
+		return xdp_adjust_head(ctx, 0 - len_diff);
```
Comment (pchaigno) [inline bpf/include/bpf/ctx/xdp.h]: Is this needed? You only call `ctx_adjust_hroom` from `handle_dsr_v4` and that function is only called from bpf_lxc, so the XDP implementation shouldn't be needed for `len_diff` < 0, no?

File: bpf/include/bpf/ctx/xdp.h [hunk 2]
```diff
@@ -210,7 +210,7 @@ ctx_adjust_hroom(struct xdp_md *ctx, const __s32 len_diff, const __u32 mode,
 	int ret;
 
 	build_bug_on(len_diff == 0 || len_diff >= 64);
-	build_bug_on(mode != BPF_ADJ_ROOM_NET);
+	build_bug_on((len_diff > 0) &&(mode != BPF_ADJ_ROOM_NET));
```
Comment (pchaigno) [inline bpf/include/bpf/ctx/xdp.h]: This should be in the previous commit, no?

File: bpf/lib/conntrack.h [hunk 3]
```diff
@@ -499,7 +499,9 @@ ipv4_extract_tuple(struct __ctx_buff *ctx, struct ipv4_ct_tuple *tuple,
 	tuple->nexthdr = ip4->protocol;
 
 	if (unlikely(tuple->nexthdr != IPPROTO_TCP &&
-		     tuple->nexthdr != IPPROTO_UDP))
+		     tuple->nexthdr != IPPROTO_UDP &&
+		     tuple->nexthdr != IPPROTO_IPIP))
```
Comment (pchaigno) [inline bpf/lib/conntrack.h]: Why is this change needed?

File: bpf/lib/conntrack.h [hunk 4]
```diff
@@ -615,7 +617,9 @@ ct_extract_ports4(struct __ctx_buff *ctx, int off, int dir,
 			}
 		}
 		break;
-
+	case IPPROTO_IPIP:
+		off = off + 20;
```
Comment (pchaigno) [inline bpf/lib/conntrack.h]: We need to check that the outer IP header isn't >20B (doesn't carry IPv4 options).

Also, let's use `sizeof(struct iphdr)` here.

File: bpf/lib/conntrack.h [hunk 5]
```diff
@@ -725,7 +729,26 @@ static __always_inline int ct_lookup4(const void *map,
...
+	case IPPROTO_IPIP:
+		{
+			__u8 protocol;
+			int iphdrlen = 20;
+
+			if (ctx_load_bytes(ctx, off +
+					offsetof(struct iphdr, protocol), &protocol, 1) < 0)
+				return DROP_CT_INVALID_HDR;
 
+			off = off + iphdrlen;
```
Comment (pchaigno) [inline bpf/lib/conntrack.h]: We should use `ipv4_hdrlen()` here to retrieve the actual size of the inner IP header.

File: bpf/lib/nodeport.h [hunk 6]
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
 }
 #endif /* DSR_ENCAP_MODE */
 
-static __always_inline int handle_dsr_v4(struct __ctx_buff *ctx, bool *dsr)
+static __always_inline int handle_dsr_v4(struct __ctx_buff *ctx, bool *dsr, int ct)
```
Comment (pchaigno) [inline bpf/lib/nodeport.h]: We recently introduced an enumeration for the conntrack states:
```suggestion
static __always_inline int
handle_dsr_v4(struct __ctx_buff *ctx, bool *dsr, enum ct_status ct)
```

File: bpf/lib/nodeport.h [hunk 7]
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
...
+			if (ct == CT_NEW) {
+				if (!revalidate_data(ctx, &data, &data_end, &ip4))
+					return DROP_INVALID;
+
+				l4_off = l3_off + ipv4_hdrlen(ip4);
+				if (ctx_load_bytes(ctx, l4_off +2, &dport, 2) < 0)
+					return DROP_CT_INVALID_HDR;
+
+				if (snat_v4_create_dsr(ctx, vip, dport) < 0)
+					return DROP_INVALID;
```
Comment (pchaigno) [inline bpf/lib/nodeport.h:1609]: Could you refactor this function to share this call with the `DSR_ENCAP_NONE` case? You probably will need to rework a bit the existing code in a preparatory commit.

File: bpf/lib/nodeport.h [hunk 8]
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
...
+#if DSR_ENCAP_MODE == DSR_ENCAP_IPIP
+	{
+		struct iphdr iph_inner __maybe_unused;
+		const int l3_off __maybe_unused = ETH_HLEN;
+		__be32 vip __maybe_unused;
+		__be32 podip;
+		__be32 sip __maybe_unused;
+		__be32 sum __maybe_unused;
+		__be16 dport __maybe_unused = 0;
+		int l4_off __maybe_unused;
```
Comment (pchaigno) [inline bpf/lib/nodeport.h]: All the `__maybe_unused` seem unnecessary here. Maybe they are from a previous version of the changes?

File: bpf/lib/nodeport.h [hunk 9]
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
...
+			/* replace outer iph with inner iph */
+			if (ctx_store_bytes(ctx, l3_off,
+				&iph_inner, sizeof(iph_inner), 0) < 0)
+				return DROP_WRITE_ERROR;
+
+			if (l3_csum_replace(ctx, ETH_HLEN + offsetof(struct iphdr, check),
+					    0, sum, 0) < 0)
+			    return DROP_CSUM_L3;
+
+			if (ct == CT_NEW) {
```
Comment (pchaigno) [inline bpf/lib/nodeport.h]: You set `ct` to `CT_NEW` on line 1379, so this condition is always true. What was the intent here? Why do we need to check the conntrack state in the first place?

File: bpf/lib/nodeport.h [hunk 10]
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
...
+		podip = ip4->daddr;
+
+		if (ip4->protocol == IPPROTO_IPIP) {
+			if (ctx_load_bytes(ctx, ETH_HLEN + sizeof(struct iphdr),
+					&iph_inner, sizeof(iph_inner)) < 0)
+				return DROP_INVALID;
+
+			*dsr = false;
+			/* this will remove inner iph */
+			if (ctx_adjust_hroom(ctx, -20,
```
Comment (pchaigno) [inline bpf/lib/nodeport.h]: We should use `sizeof(struct iphdr)` here.

File: bpf/lib/nodeport.h [hunk 11]
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
...
+		int l4_off __maybe_unused;
+
+		ct = 0;
+		if (!revalidate_data(ctx, &data, &data_end, &ip4))
+			return DROP_INVALID;
+		*dsr = false;
+		podip = ip4->daddr;
+
+		if (ip4->protocol == IPPROTO_IPIP) {
+			if (ctx_load_bytes(ctx, ETH_HLEN + sizeof(struct iphdr),
```
Comment (pchaigno) [inline bpf/lib/nodeport.h:1548]: We should check that the outer header isn't bigger than `sizeof(struct iphdr)`.

File: bpf/lib/nodeport.h [hunk 12]
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
...
+		__be32 podip;
+		__be32 sip __maybe_unused;
+		__be32 sum __maybe_unused;
+		__be16 dport __maybe_unused = 0;
+		int l4_off __maybe_unused;
+
+		ct = 0;
+		if (!revalidate_data(ctx, &data, &data_end, &ip4))
+			return DROP_INVALID;
+		*dsr = false;
```
Comment (pchaigno) [inline bpf/lib/nodeport.h]: This is not needed since `dsr` is already initialized to false.

File: bpf/lib/nodeport.h [hunk 13]
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
...
+		podip = ip4->daddr;
+
+		if (ip4->protocol == IPPROTO_IPIP) {
+			if (ctx_load_bytes(ctx, ETH_HLEN + sizeof(struct iphdr),
+					&iph_inner, sizeof(iph_inner)) < 0)
+				return DROP_INVALID;
+
+			*dsr = false;
```
Comment (pchaigno) [inline bpf/lib/nodeport.h]: This is not needed since `dsr` is already initialized to false.

File: bpf/lib/nodeport.h [hunk 14]
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
...
+
+			if (ct == CT_NEW) {
+				if (!revalidate_data(ctx, &data, &data_end, &ip4))
+					return DROP_INVALID;
+
+				l4_off = l3_off + ipv4_hdrlen(ip4);
+				if (ctx_load_bytes(ctx, l4_off +2, &dport, 2) < 0)
+					return DROP_CT_INVALID_HDR;
+
+				if (snat_v4_create_dsr(ctx, vip, dport) < 0)
```
Comment (pchaigno) [inline bpf/lib/nodeport.h]: We should set `*dsr = true` before calling that function, no?

File: bpf/lib/nodeport.h [hunk 15]
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
...
+			if (l3_csum_replace(ctx, ETH_HLEN + offsetof(struct iphdr, check),
+					    0, sum, 0) < 0)
+			    return DROP_CSUM_L3;
+
+			if (ct == CT_NEW) {
+				if (!revalidate_data(ctx, &data, &data_end, &ip4))
+					return DROP_INVALID;
+
+				l4_off = l3_off + ipv4_hdrlen(ip4);
+				if (ctx_load_bytes(ctx, l4_off +2, &dport, 2) < 0)
```
Comment (pchaigno) [inline bpf/lib/nodeport.h]: Let's not hardcode the offset here.

File: bpf/lib/nodeport.h [hunk 16]
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
...
+
+				if (snat_v4_create_dsr(ctx, vip, dport) < 0)
+					return DROP_INVALID;
+			}
+		}
+	}
+
+#elif DSR_ENCAP_MODE == DSR_ENCAP_NONE
+	if (ct != CT_NEW)
+		return 0;
```
Comment (pchaigno) [inline bpf/lib/nodeport.h:1616]: Why is this check needed?

File: daemon/cmd/kube_proxy_replacement.go [hunk 17]
```diff
@@ -351,9 +351,6 @@ func initKubeProxyReplacementOptions() (bool, error) {
 
 		if option.Config.NodePortMode == option.NodePortModeDSR &&
 			option.Config.LoadBalancerDSRDispatch == option.DSRDispatchIPIP {
-			if option.Config.DatapathMode != datapathOption.DatapathModeLBOnly {
-				return false, fmt.Errorf("DSR dispatch mode %s only supported for --%s=%s", option.Config.LoadBalancerDSRDispatch, option.DatapathMode, datapathOption.DatapathModeLBOnly)
-			}
```
Comment (pchaigno) [inline daemon/cmd/kube_proxy_replacement.go]: Could you also update the documentation at https://docs.cilium.io/en/latest/gettingstarted/kubeproxy-free/#direct-server-return-dsr?

Comment (pchaigno) [inline daemon/cmd/kube_proxy_replacement.go]: We also need to extend bpf/Makefile's `LXC_OPTIONS` with new cases to cover `DSR_ENCAP_MODE == DSR_ENCAP_NONE` and `DSR_ENCAP_MODE == DSR_ENCAP_IPIP`. See `XDP_OPTIONS` for an example.

**Review by pchaigno — CHANGES_REQUESTED** [review]
Thanks for your contribution! Being able to use IPIP with DSR without needing the LB-only mode would be great!

I reviewed this with @YutaroHayakawa. Please see our comments below. In addition to those, could you:
1. Update the release note text in the pull request description? That will be integrated to the v1.12 release notes so it needs to describe the user impact. One sentence is enough.
2. Add an end-to-end test. You can find an example of an existing end-to-end test for XDP+DSR at https://github.com/cilium/cilium/blob/bcb2db046b49dede753f91799552f31ec687f7e0/test/k8sT/Services.go#L1531-L1541.

Don't hesitate to ask on Slack if you need help with any of the suggested changes!

File: bpf/include/bpf/ctx/xdp.h (hunk 1 above)
Comment (Jiang1155) [reply bpf/include/bpf/ctx/xdp.h]: This is needed because in handle_dsr_v4, I added a new line to call ctx_adjust_hroom with a negative offset.

File: bpf/include/bpf/ctx/xdp.h (hunk 2 above)
Comment (Jiang1155) [reply bpf/include/bpf/ctx/xdp.h]: yes. I will fix it.

File: bpf/lib/conntrack.h (hunk 3 above)
Comment (Jiang1155) [reply bpf/lib/conntrack.h]: Without this change, the IPIP packets will be dropped  before handle_dsr_v4.

File: bpf/lib/conntrack.h (hunk 4 above)
Comment (Jiang1155) [reply bpf/lib/conntrack.h]: If the outer IP header is longer than 20B, what should we do? Drop it? I think another way is to check the outer IP header length and use it to replace 20B. What do you think?

I will use sizeof(struct iphdr), thanks.

File: bpf/lib/conntrack.h (hunk 5 above)
Comment (Jiang1155) [reply bpf/lib/conntrack.h]: sure.

File: bpf/lib/nodeport.h (hunk 9 above)
Comment (Jiang1155) [reply bpf/lib/nodeport.h]: I should remove line 1379. The intention is to pass CT status into this function, then only create a nat entry for the new connections.

File: bpf/lib/nodeport.h (hunk 10 above)
Comment (Jiang1155) [reply bpf/lib/nodeport.h]: ok

File: bpf/lib/nodeport.h (hunk 13 above)
Comment (Jiang1155) [reply bpf/lib/nodeport.h]: agree

File: bpf/lib/nodeport.h (hunk 14 above)
Comment (Jiang1155) [reply bpf/lib/nodeport.h]: I think you are right. Will fix it.

File: bpf/lib/nodeport.h (hunk 12 above)
Comment (Jiang1155) [reply bpf/lib/nodeport.h]: agree

File: bpf/include/bpf/ctx/xdp.h (hunk 1 above)
Comment (pchaigno) [reply bpf/include/bpf/ctx/xdp.h]: Ok, but `handle_dsr_v4` is only called from `bpf_lxc.c` which is never loaded in an XDP context (only tc). So I don't understand why changes to this limited XDP implementation are needed. `handle_dsr_v4` will use the tc implementation of `ctx_adjust_hroom` which simply relies on the `bpf_skb_adjust_room` helper.

File: bpf/lib/conntrack.h (hunk 3 above)
Comment (pchaigno) [reply bpf/lib/conntrack.h]: Did you test that? I see `ipv4_extract_tuple` is called from `extract_tuple4`, which is itself called only from `ctx_redirect_to_proxy_first`. `ctx_redirect_to_proxy_first` is called only from `bpf_host`'s `to-host` section, only in case of a packet going to the proxy. Am I missing something?

File: bpf/include/bpf/ctx/xdp.h (hunk 1 above)
Comment (Jiang1155) [reply bpf/include/bpf/ctx/xdp.h]: I got your point now. I will remove this code and test it again. Thanks.

File: bpf/lib/conntrack.h (hunk 3 above)
Comment (Jiang1155) [reply bpf/lib/conntrack.h]: I see. I did not test this change alone, but tested it with another change. I will remove it and test again. Thanks.
btw, I saw other comments about adding new test, updating doc, refactoring a function etc. I will try to address all of them.

Comment (pchaigno) [reply bpf/lib/conntrack.h]: Awesome! Don't hesitate to ask for help if needed. I'm pchaigno on Slack as well.

**Comment by pchaigno** [comment]
I'm converting to draft while the review comments are addressed, to avoid notifications for others. Don't hesitate to ask for help if needed.

File: bpf/include/bpf/ctx/xdp.h (hunk 1 above)
Comment (Jiang1155) [reply bpf/include/bpf/ctx/xdp.h]: Sorry for the late reply. I was fixing some other serious bugs before this one. I just tried to remove it and it is like what you said.  I will revert it in the next version. Thanks.

File: bpf/lib/nodeport.h (hunk 7 above)
Comment (Jiang1155) [reply bpf/lib/nodeport.h:1609]: I modified this part in the newer version. But did not share the call to snat_v4_create_dsr(). With the new logic, to share the call is a little bit complicated. We can chat more if you still think it is better to share the call. Thanks.

File: bpf/lib/nodeport.h (hunk 8 above)
Comment (Jiang1155) [reply bpf/lib/nodeport.h]: fixed. thanks

File: bpf/lib/nodeport.h (hunk 6 above)
Comment (Jiang1155) [reply bpf/lib/nodeport.h]: fixed

File: bpf/lib/nodeport.h (hunk 11 above)
Comment (Jiang1155) [reply bpf/lib/nodeport.h:1548]: Fixed

File: bpf/lib/nodeport.h (hunk 16 above)
Comment (Jiang1155) [reply bpf/lib/nodeport.h:1616]: This check is needed to make old DSR option logic the same as before.  Before the patch, handle_dsr_v4 is only called when CT is new ( and reopened in later version ).

File: daemon/cmd/kube_proxy_replacement.go (hunk 17 above)
Comment (Jiang1155) [reply daemon/cmd/kube_proxy_replacement.go]: updated

Comment (Jiang1155) [reply daemon/cmd/kube_proxy_replacement.go]: In the latest version, I added some MACROs in XDP_OPTIONS in makefile. Not sure if that is good enough or not. For now, this new ipip mode requires XDP.

File: Documentation/gettingstarted/kubeproxy-free.rst [hunk 18]
```diff
@@ -497,6 +498,42 @@ mode would look as follows:
         --set k8sServiceHost=REPLACE_WITH_API_SERVER_IP \\
         --set k8sServicePort=REPLACE_WITH_API_SERVER_PORT
 
+
+.. _DSR mode with IPIP:
+
+Direct Server Return (DSR) with IPIP
```
Comment (pchaigno) [inline Documentation/gettingstarted/kubeproxy-free.rst:474]: Shouldn't this section be right after the ` Direct Server Return (DSR) with IP option` section?

File: bpf/Makefile [hunk 19]
```diff
@@ -208,8 +208,10 @@ XDP_OPTIONS = $(LB_OPTIONS) \
 	-DENABLE_NODEPORT_ACCELERATION:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_DSR:-DENABLE_DSR_HYBRID: \
 	-DENABLE_NODEPORT_ACCELERATION:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_NODEPORT:-DENABLE_DSR:-DDSR_ENCAP_MODE:-DDSR_ENCAP_NONE:-DDSR_ENCAP_IPIP=2 \
 	-DENABLE_NODEPORT_ACCELERATION:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_NODEPORT:-DENABLE_DSR:-DDSR_ENCAP_MODE:-DDSR_ENCAP_IPIP:-DDSR_ENCAP_NONE=2 \
+	-DENABLE_NODEPORT_ACCELERATION:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_NODEPORT:-DENABLE_DSR:-DDSR_ENCAP_MODE:-DDSR_ENCAP_IPIP_CNI:-DDSR_ENCAP_NONE=3 \
```
Comment (pchaigno) [inline bpf/Makefile]: Why 3?

File: bpf/lib/common.h [hunk 20]
```diff
@@ -609,9 +609,11 @@ enum {
 	CB_POLICY,
 #define	CB_ADDR_V6_2		CB_POLICY	/* Alias, non-overlapping */
 #define	CB_BACKEND_ID		CB_POLICY	/* Alias, non-overlapping */
+#define CB_SRC_PORT		CB_POLICY	/* Alias, non-overlapping */
```
Comment (pchaigno) [inline bpf/lib/common.h]: Watch out for the mix of tab and space indentations here.

File: bpf/bpf_lxc.c [hunk 21]
```diff
@@ -1602,18 +1602,31 @@ ipv4_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label, enum ct_status
...
 
 #ifdef ENABLE_NODEPORT
-	if (ret == CT_NEW || ret == CT_REOPENED) {
+	{
 		bool dsr = false;
 # ifdef ENABLE_DSR
 		int ret2;
 
-		ret2 = handle_dsr_v4(ctx, &dsr);
+		ret2 = handle_dsr_v4(ctx, &dsr, (enum ct_status)ret);
```
Comment (pchaigno) [inline bpf/bpf_lxc.c]:
```suggestion
		ret2 = handle_dsr_v4(ctx, &dsr, *ct_status);
```

File: bpf/bpf_lxc.c [hunk 22]
```diff
@@ -1602,18 +1602,31 @@ ipv4_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label, enum ct_status
 #endif /* ENABLE_PER_PACKET_LB && !DISABLE_LOOPBACK_LB */
 
 #ifdef ENABLE_NODEPORT
-	if (ret == CT_NEW || ret == CT_REOPENED) {
```
Comment (pchaigno) [inline bpf/bpf_lxc.c:1791]: Why is this change required? Won't it impact other options besides the new one?

File: bpf/bpf_lxc.c [hunk 23]
```diff
@@ -1602,18 +1602,31 @@ ipv4_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label, enum ct_status
...
+
+#if DSR_ENCAP_MODE == DSR_ENCAP_IPIP_CNI
+		if (!revalidate_data(ctx, &data, &data_end, &ip4))
+			return DROP_INVALID;
+		tuple.nexthdr = ip4->protocol;
+		tuple.daddr = ip4->daddr;
+		tuple.saddr = ip4->saddr;
+
+		ret = ct_lookup4(get_ct_map4(&tuple), &tuple, ctx, l4_off, CT_INGRESS, &ct_state,
+				 &monitor);
```
Comment (pchaigno) [inline bpf/bpf_lxc.c:1813]: Why do we need to lookup the connection state again here when we already performed a lookup on line 1529?

File: bpf/lib/nodeport.h [hunk 24]
```diff
@@ -523,6 +523,9 @@ int tail_nodeport_ipv6_dsr(struct __ctx_buff *ctx)
 #elif DSR_ENCAP_MODE == DSR_ENCAP_NONE
 	port = (__u16)ctx_load_meta(ctx, CB_PORT);
 	ret = dsr_set_ext6(ctx, ip6, &addr, port, &ohead);
+#elif DSR_ENCAP_MODE == DSR_ENCAP_IPIP_CNI
+	/*To do, add support for ipv6 */
```
Comment (pchaigno) [inline bpf/lib/nodeport.h:527]: That limitation should be documented.

File: bpf/lib/nodeport.h [hunk 25]
```diff
@@ -753,7 +756,11 @@ static __always_inline int nodeport_lb6(struct __ctx_buff *ctx,
 
 	svc = lb6_lookup_service(&key, false);
 	if (svc) {
-		const bool skip_l3_xlate = DSR_ENCAP_MODE == DSR_ENCAP_IPIP;
+#if DSR_ENCAP_MODE == DSR_ENCAP_IPIP
+		const bool skip_l3_xlate = true;
+#else
+		const bool skip_l3_xlate = false;
+#endif
```
Comment (pchaigno) [inline bpf/lib/nodeport.h:763]:
```suggestion
		const bool skip_l3_xlate = DSR_ENCAP_MODE == DSR_ENCAP_IPIP;
```

File: bpf/Makefile (hunk 19 above)
Comment (pchaigno) [reply bpf/Makefile]: We should also update the configs under `bpf/complexity-tests/` to cover this new option. It's likely to have a complexity higher than the current DSR options.

File: bpf/lib/conntrack.h (hunk 4 above)
Comment (pchaigno) [reply bpf/lib/conntrack.h]: > I think another way is to check the outer IP header length and use it to replace 20B. What do you think?

Yes, I think that's the right approach.

**Review by pchaigno — CHANGES_REQUESTED** [review]
I didn't review the core of the logic, but have a few other minor comments that may be worth addressing in the next iteration.

**Review by brb — CHANGES_REQUESTED** [review]
Thanks for the PR! I think you need to mention in the docs and also implement a check in `pkg/service/service.go` to check that a backend port matches a service port, as with the IP-in-IP encapsulation we cannot have L4 different ports.

**Comment by Jiang1155** [comment]
> Your branch is based on a 2 weeks old version of `master`. I think it would be best to rebase before triggering the end-to-end tests. That may also fix the failing test (https://github.com/cilium/cilium/runs/6779786562?check_suite_focus=true).
> 
> Please also check the minor warning in checkpatch.

Sure.  I will update to the latest master. As for checkpatch, I run "make -C bpf checkpatch" and the only warning is the following one:
WARNING:PREFER_FALLTHROUGH: Prefer 'fallthrough;' over fallthrough comment
#98: FILE: bpf/lib/conntrack.h:610:
+               /* fallthrough; */
That does not apply here because ebpf does not have fallthrough keywords. So I think I will just ignore it. Did I miss anything that need to be fixed?

btw, I saw you other comments about style like mixing space and tabs. I will fix those. Not sure why it does not show up in the make commands.

File: Documentation/gettingstarted/kubeproxy-free.rst (hunk 18 above)
Comment (Jiang1155) [reply Documentation/gettingstarted/kubeproxy-free.rst:474]: Sure. Will do.

File: bpf/Makefile (hunk 19 above)
Comment (Jiang1155) [reply bpf/Makefile]: I will change that to 2. I was a little bit confused when I changing it to 3.  Will also update complexity-tests. I think just need to update bpf_lxc.txt file, right?

File: bpf/lib/common.h (hunk 20 above)
Comment (Jiang1155) [reply bpf/lib/common.h]: will do.

File: bpf/bpf_lxc.c (hunk 21 above)
Comment (Jiang1155) [reply bpf/bpf_lxc.c]: fixed

File: bpf/bpf_lxc.c (hunk 22 above)
Comment (Jiang1155) [reply bpf/bpf_lxc.c:1791]: I moved CT== new or reopened check inside the handle_dsr_v4() function, so it will not impact old DSR with IP option mode. 

But this is a good question, it reminds me that maybe I can optimize current code a little bit.
My original thought is that we will do IPIP encap for all tcp packets (not only syn packets as used in ip option mode).  But now I think maybe I don't need to do that. I can do something similar as IP option and only use IPIP encap for syn packets.  I will give it a try.

File: bpf/bpf_lxc.c (hunk 23 above)
Comment (Jiang1155) [reply bpf/bpf_lxc.c:1813]: The tuple changed. The lookup at line 1529 is looking for outer tuples. After handle_dsr_v4 on line 1610, the nexthdr and saddr changed. If I don't do lookup here, it still works. But there will be duplicate ct entries for a single connection.

File: bpf/lib/nodeport.h (hunk 24 above)
Comment (Jiang1155) [reply bpf/lib/nodeport.h:527]: Sure. will do.

File: bpf/lib/nodeport.h (hunk 25 above)
Comment (Jiang1155) [reply bpf/lib/nodeport.h:763]: I first attempt is the same as you suggested. But I got following error and had a hard time to figure out why.

/work/go/src/github.com/cilium/cilium/bpf/lib/nodeport.h:766:48: error: use of undeclared identifier 'DSR_ENCAP_IPIP'
                const bool skip_l3_xlate = DSR_ENCAP_MODE == DSR_ENCAP_IPIP;

So I used current way to make the compiler happy.

**Comment by Jiang1155** [comment]
> Thanks for the PR! I think you need to mention in the docs and also implement a check in `pkg/service/service.go` to check that a backend port matches a service port, as with the IP-in-IP encapsulation we cannot have L4 different ports.

With pure IP-in-IP, we cannot have different L4 ports. But in my implementation, I put the original port into the inner IP options for syn packets. So now a  TCP syn packet has two IP headers. Inner IP header with IP options as the old way. And a normal Outer IP header. So the L4 ports can be different. I do this to support nodeport case. Otherwise, nodeport will be broken (or need additional changes).

**Comment by brb** [comment]
> But in my implementation, I put the original port into the inner IP options for syn packets. So now a TCP syn packet has two IP headers. Inner IP header with IP options as the old way.

I think it would make sense to add an option to opt out from adding original port into the inner hdr option, and to ignore provisioning of services which port != target port. Otherwise, we risk such SYN packets to be dropped by intermediate routers which do not support IP options and can parse the inner hdr. /cc @borkmann

**Comment by brb** [comment]
> Or you mean add another option to use IPIP in CNI mode without using inner IP option?

Yes.

**Comment by Jiang1155** [comment]
> > Or you mean add another option to use IPIP in CNI mode without using inner IP option?
> 
> Yes.

I see. I have some question about how to support nodeport in this case.  Let's assume there are two nodes, and 2nd node is the real server who has pod to serve the request. In old DSR case, when first node gets a packet, it will translate both IP and port, and then forward to the 2nd node.  In ipip cni without ip option case, how do we support node port? 

One way I can think of is to split IP and port translation. On the first node, it only translate dest ip to the node ip of the 2nd node. And keep the original dest IP in inner ip header. Then on the 2nd node, it first translates outer dest ip and dest port to the pod ip and service pod. And then save inner dest ip somewhere and use it as the source ip later on. I guess this is doable, but is a little bit complicate. Is there a better/cleaner solution?

Also, could I merge the current version first ( if no other issues) and add that new option in a follow up pr? I can open an github issue to track the new option.

**Comment by brb** [comment]
> In ipip cni without ip option case, how do we support node port?

We would not support. Just log a warning in cilium-agent and ignore such a service (svc port != ep port).

**Comment by Jiang1155** [comment]
> > In ipip cni without ip option case, how do we support node port?
> 
> We would not support. Just log a warning in cilium-agent and ignore such a service (svc port != ep port).

OK. That will be much simpler.

**Comment by Jiang1155** [comment]
> > But in my implementation, I put the original port into the inner IP options for syn packets. So now a TCP syn packet has two IP headers. Inner IP header with IP options as the old way.
> 
> I think it would make sense to add an option to opt out from adding original port into the inner hdr option, and to ignore provisioning of services which port != target port. Otherwise, we risk such SYN packets to be dropped by intermediate routers which do not support IP options and can parse the inner hdr. /cc @borkmann

I think maybe we can just reuse current ipip option for this purpose. Current ipip already did ipip encap as we needed. Only two things are left is 1) it does not work other than in lb-only mode 2) it does not do ipip decap.  I can modify the code for this option to work in cni mode and do decap. I think for those users who already using ipip in lb-only mode, this will not break their system after upgrading?
btw, my first version of pr is almost the same as this one. :) Just missing warning messages.

**Comment by Jiang1155** [comment]
> > In ipip cni without ip option case, how do we support node port?
> 
> We would not support. Just log a warning in cilium-agent and ignore such a service (svc port != ep port).

@brb  How do we log a warning? Right now, I check the service type in nodeport.h->nodeport_lb4() and return DROP_ENCAP_PROHIBITED if it is a nodeport service. I guess other services ( external-ip, load-balancer are OK).  I add check here because I think it is per packet check.  If that is OK, I was thinking using cilium_dbg to log a message. But wondering if I use existing DBG types or add a new one. Any comments? Thanks.

**Comment by Jiang1155** [comment]
> > How do we log a warning?
> 
> I suggest to log the warning from the control plane, i.e., when provisioning the service in `pkg/service/service.go`.

I still have some questions about high level how this will be used. If I run "kubectl describe svc xxx", I saw "Port" and "TargetPort". You mentioned svc port and ep point. Is svc port the same as Ports and ep port the same as TargetPort?  Also, in pkg/service/service.go, I did not find svc port or ep point. There some ports like frontend ports, backend ports. Are they the same?
In addition, k8s has 4 service types, cluster IP, nodeport, load balancer and external names. How the port rule applies to those services? 
For cluster IP, since it is only internal service, my current implementation does not cover it. Do we need to do IPIP for cluster IP? 
For nodeport,  there are two cases: 1) the nodeport does not have an external IP. In this case, my current code does not support it.  2) the nodeport has an external IP, and my code will allow a client to connect to the service via external IP and the original port ( not the node port ).  Is this behavior good enough? If so, I am not sure if we can check for pkg/service/servic.go or not. The user could first create a nodeport service and later on use patch command to add an external IP. How do we handle this case? 
For load balancer, do we want to support IPIP? I don't have an environment to test this service.
For external names, I guess we can skip it too?

**Comment by brb** [comment]
> You mentioned svc port and ep point. Is svc port the same as Ports and ep port the same as TargetPort?

Yes. In pkg/service/service.go Port = Frontend Port, while TargetPort = Backend Port.

> How the port rule applies to those services?

The IPIP is going to be used only for N/S traffic. The warning should be logged for the following service types if any of its backend port doesn't match the service port:
- NodePort
- LoadBalancer
- externalIPs
- ClusterIP if `--bpf-lb-external-clusterip=true`.

The logging (and then ignoring provisioning) could happen https://github.com/cilium/cilium/blob/a13b066eef546b5aa18885aaa76469c80cb1bb1d/pkg/service/service.go#L517.

**Comment by brb** [comment]
> With IPIP ( and no ip option inside inner IP), how do we pass the port info 30088 to the real node?

It will be stored in the L4 (e.g., TCP) hdr.
